src/array_slice.cpp

```diff
@@ -564,10 +549,36 @@ realnum *collapse_array(realnum *array, int *rank, size_t dims[3], direction dir
       reduced_stride[reduced_stride[0] != 0 ? 0 : 1] = reduced_dims[1];
   }
 
+  /*--------------------------------------------------------------*/
+  reduced_grid_size = reduced_dims[0] * (reduced_rank == 2 ? reduced_dims[1] : 1);
```

Comment:
This doesn't seem to work if `reduced_rank == 3`?

It seems easier to just initialize `reduced_dims[0] = reduced_dims[1] = reduced_dims[2] = 1` and
then the caller can compute `reduced_grid_size = reduced_dims[0] * reduced_dims[1] * reduced_dims[2]` if they want.

Then don't have a `reduced_grid_size` argument here.